### DISCUSSION TITLE
Any way to modify the request path in middleware?

### DISCUSSION THREAD

**Discussion by kornysietsma** [honojs/hono#2642] [Q&A] [answered]
I am trying to replicate a legacy API which had optional versions in paths, so `GET /v1/foo` actually called `GET /foo` under the covers.

I was hoping to add some middleware that just intercepted paths starting with `/v1` and removed the prefix (or re-directed somehow to the correct route) - but it seems that modifying the `ctx.req.path` does nothing.

Is this possible?  I've seen examples that clone `ctx.req.raw` and make a brand new request - that feels a bit ugly, and I'm not sure that Hono will use the new request's path for routing anyway.

Any suggestions?  (for added complexity we are using the zod-openapi middleware so I can't just give our routes both `/foo` and `/v1/foo` paths easily.  And we'd like to decrement the `/v1` rubbish anyway, so I'd prefer to not have it in our openapi schema)

**Comment by ghost** [comment]
How about using `redirect()` to handle access to legacy API endpoints?

```ts
import { Hono } from 'hono'

const app = new Hono()

app.get('/api/hello', (c) => {
  return c.text('api')
})

app.use('/api/v1/*', async (c) => {
  const newPath = c.req.path.replace(/^\/api\/v1/, '/api');
  // 301 means redirect-permanently
  return c.redirect(newPath)
})
```

This avoids setting up legacy APIs one at a time.
Including the version in the API path is a safe way to manage breaking changes, so decide carefully whether to use it.

**Reply by kornysietsma** [reply, marked as answer]
sadly that only works if our external clients all respect `301 redirect` and I'm not totally confident they do.
What we decided on was to move to hono v4 and just specify multiple paths.  (and we ditched the hod-zono-openapi stuff so we can hide the `/v1` from our schema)

Our resulting code:

```typescript
  app.on('GET', ['/foo', '/v1/foo'],
```

Not as generic as a middleware for all routes - but a lot less work.
